providers/frame/polkadot_dev.ts

```diff
@@ -16,15 +30,100 @@ export class PolkadotDevProvider extends FrameBinProvider {
+        L extends number,
+        A extends T[] = [],
+      > = number extends L ? T[]
+        : L extends A["length"] ? A
+        : ArrayOfLength<T, L, [...A, T]>
+
+      export async function users<N extends number>(count: N): Promise<ArrayOfLength<C.Sr25519, N>>
+      export async function users(count: number): Promise<C.Sr25519[]> {
+        const response = await fetch(
+          "${url}",
+          {
+            method: "POST",
+            headers: {
+              "content-type": "application/json",
+            },
+            body: JSON.stringify({ count }),
+          },
+        )
+        if (response.status >= 400) {
+          throw new Error(await response.text())
+        }
+        const { from, to }: { from: number; to: number } = await response.json()
+        const userIds: C.Sr25519[] = []
+        for (let i = from; i < to; i++) {
+          userIds.push(C.testUser(i))
+        }
+        return userIds
+      }
+    `)
+  }
+
+  override async handle(request: Request, pathInfo: PathInfo): Promise<Response> {
+    if (request.method.toUpperCase() === "POST" && pathInfo.filePath === "test-users") {
+      try {
+        const { count } = await request.json()
+        if (!(count > 0)) {
+          throw new Error("'count' should be greater than 0")
+        }
+        const currentCount = this.#testUserCountCache[this.#getRuntime(pathInfo)] ?? 0
+        if (count + currentCount > TEST_USER_COUNT) {
+          throw new Error("Maximum test user count reached")
+        }
+        const from = currentCount
+        const to = currentCount + count
+        this.#testUserCountCache[this.#getRuntime(pathInfo)] = to
+        return new Response(
+          JSON.stringify({ from, to }),
+          {
+            status: 200,
+            headers: {
+              "content-type": "application/json",
+            },
+          },
+        )
+      } catch (error) {
+        console.log("Error", request.url, error)
+        const message = error instanceof Error
+          ? error.message
+          : "Unknown error"
+        return new Response(message, { status: 400 })
+      }
+    }
+    return super.handle(request, pathInfo)
+  }
+
   async launch(pathInfo: PathInfo) {
-    const runtimeName = pathInfo.target
+    const runtimeName = this.#getRuntime(pathInfo)
     $.assert($devRuntimeName, runtimeName)
     const port = getAvailable()
-    const args: string[] = ["--dev", "--ws-port", port.toString()]
-    if (runtimeName !== "polkadot") args.push(`--force-${runtimeName}`)
+    const chainSpec = await createCustomChainSpec({
+      binary: this.bin,
+      chain: `${runtimeName}-dev`,
+      testUserAccountProps: {
+        networkPrefix: DEV_RUNTIME_PREFIXES[runtimeName],
+        count: TEST_USER_COUNT,
+      },
+    })
+    const args: string[] = ["--tmp", "--alice", "--ws-port", port.toString(), "--chain", chainSpec]
     await this.runBin(args)
     return port
   }
+
+  #getRuntime(pathInfo: PathInfo) {
```

Comment:
Why is extracting this necessary?

Reply:
I agree that the current extraction is unnecessary, but I think extracting the assert (as per my review comment on this) is warranted.

Reply:
it is to normalize the user count cache key 
as suggested previously, I'll remove the normailization